src/new_menu_helpers.c

```diff
@@ -0,0 +1,765 @@
+        {
+            for (i = 0; i < gUnknown_203AB5C; i++)
+            {
+                FREE_AND_SET_NULL(gUnknown_203AB60[i]);
+            }
+            gUnknown_203AB5C = 0;
+        }
+        return FALSE;
+    }
+    else
+    {
+        return TRUE;
+    }
+}
+
+void *DecompressAndCopyTileDataToVram(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    if (gUnknown_203AB5C < NELEMS(gUnknown_203AB60))
+    {
+        void *ptr = MallocAndDecompress(src, &sizeOut);
+        if (!size)
+            size = sizeOut;
+        if (ptr)
+        {
+            CopyDecompressedTileDataToVram(bgId, ptr, size, offset, mode);
+            gUnknown_203AB60[gUnknown_203AB5C++] = ptr;
+        }
+        return ptr;
+    }
+    return NULL;
+}
+
+void *DecompressAndCopyTileDataToVram2(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    if (gUnknown_203AB5C < NELEMS(gUnknown_203AB60))
+    {
+        void *ptr = MallocAndDecompress(src, &sizeOut);
+        if (sizeOut > size)
+            sizeOut = size;
+        if (ptr)
+        {
+            CopyDecompressedTileDataToVram(bgId, ptr, sizeOut, offset, mode);
+            gUnknown_203AB60[gUnknown_203AB5C++] = ptr;
+        }
+        return ptr;
+    }
+    return NULL;
+}
+
+void DecompressAndLoadBgGfxUsingHeap(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    void *ptr = MallocAndDecompress(src, &sizeOut);
+    if (!size)
+        size = sizeOut;
+    if (ptr)
+    {
+        u8 taskId = CreateTask(TaskFreeBufAfterCopyingTileDataToVram, 0);
+        gTasks[taskId].data[0] = CopyDecompressedTileDataToVram(bgId, ptr, size, offset, mode);
+        SetWordTaskArg(taskId, 1, (u32)ptr);
+    }
+}
+
+void DecompressAndLoadBgGfxUsingHeap2(u8 bgId, const void *src, u32 size, u16 offset, u8 mode)
+{
+    u32 sizeOut;
+    void *ptr = MallocAndDecompress(src, &sizeOut);
+    if (sizeOut > size)
+        sizeOut = size;
+    if (ptr)
+    {
+        u8 taskId = CreateTask(TaskFreeBufAfterCopyingTileDataToVram, 0);
+        gTasks[taskId].data[0] = CopyDecompressedTileDataToVram(bgId, ptr, sizeOut, offset, mode);
+        SetWordTaskArg(taskId, 1, (u32)ptr);
+    }
+}
+
+static void TaskFreeBufAfterCopyingTileDataToVram(u8 taskId)
+{
+    if (!CheckForSpaceForDma3Request(gTasks[taskId].data[0]))
+    {
+        Free((void *)GetWordTaskArg(taskId, 1));
+        DestroyTask(taskId);
+    }
+}
+
+void *MallocAndDecompress(const void *src, u32 *size)
+{
+    void *ptr;
+    u8 *sizeAsBytes = (u8 *)size;
+    const u8 *srcAsBytes = src;
+
+    sizeAsBytes[0] = srcAsBytes[1];
+    sizeAsBytes[1] = srcAsBytes[2];
+    sizeAsBytes[2] = srcAsBytes[3];
+    sizeAsBytes[3] = 0;
+
+    ptr = Alloc(*size);
+    if (ptr)
+        LZ77UnCompWram(src, ptr);
+    return ptr;
+}
+
+static u16 CopyDecompressedTileDataToVram(u8 bgId, const void *src, u16 size, u16 offset, u8 mode)
+{
+    switch (mode)
+    { // different to EM
```

Comment:
what does this comment mean?

Reply:
The function logic got changed in EM. If you regard it as confusing I can delete all this kind of comment

Reply:
it's not necessary. if anything, the emerald version should have a "different from fr/lg" comment as that game released after frlg